lib/std/smartptrs.nim

```diff
@@ -0,0 +1,222 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
+
+
+when defined(nimExperimentalSmartptrs) or defined(nimdoc):
+  import std/isolation
+
+
+  type
+    Deleter*[T] = proc (val: var T)
+    UniquePtr*[T] = object
+      ## Non copyable pointer to a value of type `T` with exclusive ownership.
+      val: ptr T
+      deleter: Deleter[T]
+
+  proc `=destroy`*[T](p: var UniquePtr[T]) =
+    if p.val != nil:
+      if p.deleter != nil:
+        p.deleter(p.val[])
+      else:
+        `=destroy`(p.val[])
+      when compileOption("threads"):
+        deallocShared(p.val)
+      else:
+        dealloc(p.val)
+
+  proc `=copy`*[T](dest: var UniquePtr[T], src: UniquePtr[T]) {.error.}
+    ## The copy operation is disallowed for `UniquePtr`, it
+    ## can only be moved.
+
+  proc newUniquePtr[T](val: sink Isolated[T], deleter: Deleter[T] = nil): UniquePtr[T] {.nodestroy.} =
```

Comment:
this does not appear to support stateful deleters, which is something you can do in c++.

I also observe that space is used for the deleter even if you don't specify a non-default one